airbyte-integrations/connectors/source-dbt-cloud/source_dbt_cloud/source.py

```diff
@@ -0,0 +1,329 @@
+#
+# MIT License
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+from requests.api import head
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+
+
+class DbtCloudStream(HttpStream, ABC):
+    """
+    This class represents a stream output by the connector.
+    """
+
+    url_base = "https://cloud.getdbt.com/api/v2/"
+
+    primary_key = None
+
+    def next_page_token(
+        self, response: requests.Response, stream_batch_count: int
+    ) -> Optional[Mapping[str, Any]]:
+        """
+        In each response given by the dbt cloud API, we're able to retrieve total number of occurences for the requested object and the number of occurences returned in this batch.
+        "pagination": {
+            "count": 10,
+            "total_count": 2012
+        }
+        Our strategy here is to sum up all batches count and enter exit condition if the sum of all batches count is equal or greater (shouldn't happen)
+        than total count as received from API response.
+        """
+        batch_count = response.json().get("extra").get("pagination").get("count")
```

Comment:
Je serais plus partant pour des safe calls, juste in case
`batch_count = response.json().get("extra", {}).get("pagination", {}).get("count")`